### PR TITLE
Fixes issue with Numba CUDA target altering max/min semantics.

### PR DESCRIPTION
**Pull request by stuartarchibald**
As title.

Closes https://github.com/holoviz/datashader/issues/932

### CODE REVIEW

**Comment by stuartarchibald** [comment]
This is entirely untested and relies on https://github.com/numba/numba/pull/5941, open question about what to do for version Numba 0.50. which has neither the old atomic min/max behaviour nor the new nanmin/nanmax functions.

File: datashader/transfer_functions/_cuda_utils.py [hunk 1]
```diff
@@ -70,13 +72,35 @@ def masked_clip_2d(data, mask, lower, upper):
...
+    def cuda_atomic_nanmax(ary, idx, val)
+        return cuda.atomic.nanmax(ary, idx, val)
+elif LooseVersion(numba.__version__) <= LooseVersion("0.49.0"):
+    @cuda.jit
+    def cuda_atomic_nanmin(ary, idx, val)
+        return cuda.atomic.min(ary, idx, val)
+
+    @cuda.jit
+    def cuda_atomic_nanmax(ary, idx, val)
+        return cuda.atomic.max(ary, idx, val)
```
Comment (jbednar) [inline datashader/transfer_functions/_cuda_utils.py:91]: Are the `def`s really needed here?  Would it not be sufficient to do `cuda_atomic_nanmin = cuda.atomic.nanmin` in the first `if` and `cuda_atomic_nanmin = cuda.atomic.min` in the second?

File: datashader/transfer_functions/_cuda_utils.py [hunk 2]
```diff
@@ -70,13 +72,35 @@ def masked_clip_2d(data, mask, lower, upper):
...
+else:
+    # WHAT DO YOU WANT TO DO HERE?
```
Comment (jbednar) [inline datashader/transfer_functions/_cuda_utils.py:93]:
```suggestion
    assert("Datashader's CUDA support requires numba!=0.50.0")
```

File: datashader/transfer_functions/_cuda_utils.py [hunk 3]
```diff
@@ -70,13 +72,35 @@ def masked_clip_2d(data, mask, lower, upper):
...
+
 @cuda.jit
 def masked_clip_2d_kernel(data, mask, lower, upper):
     i, j = cuda.grid(2)
     maxi, maxj = data.shape
     if i >= 0 and i < maxi and j >= 0 and j < maxj and not mask[i, j]:
-        cuda.atomic.max(data, (i, j), lower)
-        cuda.atomic.min(data, (i, j), upper)
+        cuda_atomic_max(data, (i, j), lower)
+        cuda_atomic_min(data, (i, j), upper)
```
Comment (jbednar) [inline datashader/transfer_functions/_cuda_utils.py:103]:
```suggestion
        cuda_atomic_nanmax(data, (i, j), lower)
        cuda_atomic_nanmin(data, (i, j), upper)
```

File: datashader/transfer_functions/_cuda_utils.py (hunk 1 above)
Comment (stuartarchibald) [reply datashader/transfer_functions/_cuda_utils.py:91]: Unfortunately I think they are (and these may need to be device functions too), the `cuda.atomic.max/min` are stubs that are basically a token known to the compiler, if they appear as globals I don't think the compiler will know what to do with them.

**Comment by stuartarchibald** [comment]
Thanks for the review @jbednar, once 0.51 is out I'll revisit (or someone else can, I have no particular attachment to the patch, feel free to push to my branch :)) with view to test this and make it actually work opposed to it just being a guess!

File: datashader/transfer_functions/_cuda_utils.py (hunk 1 above)
Comment (jbednar) [reply datashader/transfer_functions/_cuda_utils.py:91]: Ok, thanks.

**Comment by dHannasch** [comment]
> Thanks! I haven't tested it, but hopefully @jonmmease and/or @philippjfr can.

If anyone's up for testing this, I've fixed the picky little errors at https://github.com/holoviz/datashader/pull/940. (See https://github.com/stuartarchibald/datashader/pull/1.)
Note however that https://github.com/holoviz/datashader/pull/940 has one more change that might be important. @philippjfr [asked](https://github.com/holoviz/datashader/pull/940#discussion_r453246153) why we couldn't just import `numba.cuda.atomic.nanmin` and `numba.cuda.atomic.nanmax` directly (that is, why do we need to wrap them in `@cuda.jit def`s?), and I didn't have an answer, so I tried splitting the difference and using them directly in reductions.py `_append_cuda`, but using the `@cuda.jit` wrappers in `masked_clip_2d_kernel`. So if something goes wonky in `_append_cuda` but not in `masked_clip_2d_kernel`, blame me, not @stuartarchibald.

Also, if you're going to manually test, best to just test Python 3.7 for now. There are currently unrelated problems with Dask that make the test suite fail on 2.7 and 3.6.

**Comment by stuartarchibald** [comment]
xref https://github.com/numba/numba/pull/6020 which is @gmarkall carrying on with some extra details on top of the original PR on the Numba side.

**Comment by AjayThorve** [comment]
@dHannasch I tested this with numba 0.51(https://github.com/numba/numba/releases/tag/0.51.0rc1), `python 3.7.8`.

The version check seems to work as well. Just a couple of fixes required

**Comment by philippjfr** [comment]
I'd like to go ahead and merge and then fix things up in a separate PR. Any objections?

**Comment by AjayThorve** [comment]
> I'd like to go ahead and merge and then fix things up in a separate PR. Any objections?

@philippjfr that works

**Pull request merged**